### ISSUE TITLE
Eval bug: NextN (PR #22673) Vulkan path produces ~0.7% accept rate + garbled output on RDNA3+RADV, while HIP/ROCm works correctly (81%)

### ISSUE THREAD

**Issue by kimtk44** [ggml-org/llama.cpp#23088]
## Context

This is a finding from testing PR #22673 (NextN/MTP support) as bundled in the downstream fork [`AtomicBot-ai/atomic-llama-cpp-turboquant`](https://github.com/AtomicBot-ai/atomic-llama-cpp-turboquant) at HEAD `0a635dc` (PR #14 in that fork rebases on top of this PR). The fork's issue tracker is disabled, so I'm cross-posting here since the NextN implementation under review is the relevant venue. cc @am17an

Headline: **NextN draft on the Vulkan backend produces ~0.7% acceptance and garbled output on RDNA3+RADV, while the HIP/ROCm backend on the same hardware achieves 81% acceptance + the expected +35.7% tg gain** (matching the M4 Max +36% extrapolation in the fork's NEXTN.md). The bug appears localized to the Vulkan execution of the NextN draft graph; the rest of the Vulkan backend works correctly for this model when `--spec-type` is omitted.

## Name and Version

```
$ ~/llm/bin-atomic/llama-server --version
version: 1 (0a635dc)
built with GNU 13.3.0 for Linux x86_64
```

build_info from server log: `b1-0a635dc`

## Operating systems

- Linux (Ubuntu 24.04, kernel 6.8.0-111-generic)

## GGML backends

- **Vulkan** (broken)
- **HIP** (works — reference comparison in this report)

## Hardware

- 2x AMD Radeon RX 7900 XTX (RDNA3, gfx1100)
- Mesa 25.2.8 (RADV NAVI31)
- Vulkan 1.3.275 headers (libvulkan.so.1.3.275)
- ROCm 6.4.0 (hipcc, gfx1100)
- AMD Ryzen 9 7900 (12 cores) + 64 GB DDR5

## Models

- `AtomicChat/Qwen3.6-35B-A3B-UDT-MTP-GGUF` — `Qwen3.6-35B-A3B-UDT-Q4_K_XL_MTP.gguf`
- HF lfs oid: `56aeaf6a72cb2a538a7356941fe2983cff7c5efe91cac866336f33d144913cd8`
- Total size: 22,187,497,792 B (≈20.7 GiB)
- Architecture: `qwen35moe`, `nextn_predict_layers=1`, block_count=41 (40 main + 1 NextN)
- HF: https://huggingface.co/AtomicChat/Qwen3.6-35B-A3B-UDT-MTP-GGUF
- Also reproduced with `unsloth/Qwen3.6-35B-A3B-MTP-GGUF` — `UD-Q4_K_M` (sha256 `0b21525e972670ed59e1812e170b27c26355381f0656ecc4e25617ece7dac58b`, 22.66 GB) to rule out the UDT-mask hypothesis.

### Summary

With the same UDT GGUF and same `--spec-type nextn` invocation pattern (shared-model path, target == draft), the **Vulkan backend** produces a NextN draft acceptance rate of ~0.7% and garbled multilingual token output, while the **HIP/ROCm backend** on the same hardware produces 81% acceptance with coherent output and the expected +36% throughput uplift over the NextN-OFF baseline.

NextN initialization succeeds on both backends (`NextN draft: shared-model path (target qwen35moe, no second mmap; nextn_predict_layers=1)` followed by `NextN draft context registered` and `speculative decoding context initialized`). The difference is in the per-token draft prediction signal.

### Reproduction — Vulkan (broken)

Build:

```bash
git clone --depth 1 https://github.com/AtomicBot-ai/atomic-llama-cpp-turboquant
cd atomic-llama-cpp-turboquant
cmake -B build -DCMAKE_BUILD_TYPE=Release -DGGML_VULKAN=ON \
  -DLLAMA_BUILD_TESTS=OFF -DLLAMA_BUILD_EXAMPLES=OFF -DLLAMA_BUILD_TOOLS=ON
cmake --build build --config Release -j 8 -- llama-server
```

Run:

```bash
./build/bin/llama-server \
  --model /path/to/Qwen3.6-35B-A3B-UDT-Q4_K_XL_MTP.gguf \
  --model-draft /path/to/Qwen3.6-35B-A3B-UDT-Q4_K_XL_MTP.gguf \
  --spec-type nextn --draft-max 2 --draft-min 1 \
  -c 8192 -ngl 99 -ngld 99 -fa on \
  -ctk q8_0 -ctv q8_0 -ctkd q8_0 -ctvd q8_0 \
  --device Vulkan0 --device-draft Vulkan0 \
  --host 127.0.0.1 --port 8089 \
  --parallel 1 -np 1 --cont-batching \
  --no-mmap --mlock \
  --temp 0.6 --top-p 0.95 --top-k 20 --min-p 0 \
  --jinja --log-timestamps --log-prefix --metrics --slots
```

Send a chat completion:

```bash
curl -X POST http://127.0.0.1:8089/v1/chat/completions -H "Content-Type: application/json" -d '{
  "model":"qwen3.6",
  "messages":[{"role":"user","content":"Explain in two sentences what speculative decoding is."}],
  "max_tokens":200,"temperature":0.6
}'
```

Observed (Vulkan, `--draft-max 2`):

```
draft acceptance rate = 0.00673 (    2 accepted /   297 generated)
statistics nextn: #calls(b,g,a) = 1 296 1, #gen drafts = 296, #acc drafts = 1, #gen tokens = 297, #acc tokens = 2,
  dur(b,g,a) = 116.794, 799.380, 0.003 ms
HTTP 500: Failed to parse input at pos 514: safezoneش beh again álla  Ristorante **:** dicip ...
  [multilingual garbled token soup, ~290 more tokens of similar gibberish]
```

Observed (Vulkan, `--draft-max 1`) — same pattern, rules out chain-length:

```
draft acceptance rate = 0.00680 (    1 accepted /   147 generated)
statistics nextn: #calls(b,g,a) = 1 147 1, #gen drafts = 147, #acc drafts = 1, #gen tokens = 147, #acc tokens = 1
HTTP 500: Failed to parse input at pos 22: <think>\nHere's a cogsayicer Potr}@ACHINEAsyncResulthazi spasuceToPoint xákti...
```

### Reproduction — HIP (works, reference)

Build:

```bash
export PATH=/opt/rocm/bin:$PATH
cmake -B build-hip -DCMAKE_BUILD_TYPE=Release -DGGML_HIP=ON -DGGML_HIP_ROCWMMA_FATTN=ON \
  -DAMDGPU_TARGETS=gfx1100 \
  -DCMAKE_C_COMPILER=/opt/rocm/llvm/bin/clang -DCMAKE_CXX_COMPILER=/opt/rocm/llvm/bin/clang++ \
  -DLLAMA_BUILD_TESTS=OFF -DLLAMA_BUILD_EXAMPLES=OFF -DLLAMA_BUILD_TOOLS=ON
cmake --build build-hip --config Release -j 8 -- llama-server
```

Run (same args except `--device ROCm0 --device-draft ROCm0`):

```bash
./build-hip/bin/llama-server \
  --model /path/to/Qwen3.6-35B-A3B-UDT-Q4_K_XL_MTP.gguf \
  --model-draft /path/to/Qwen3.6-35B-A3B-UDT-Q4_K_XL_MTP.gguf \
  --spec-type nextn --draft-max 2 --draft-min 1 \
  -c 8192 -ngl 99 -ngld 99 -fa on \
  -ctk q8_0 -ctv q8_0 -ctkd q8_0 -ctvd q8_0 \
  --device ROCm0 --device-draft ROCm0 \
  ...
```

Observed (HIP, `--draft-max 2`) — correct behavior:

```
draft acceptance rate = 0.81212 (  134 accepted /   165 generated)
statistics nextn: #calls(b,g,a) = 1 164 133, #gen drafts = 164, #acc drafts = 1, #gen tokens = 165, #acc tokens = 134,
  dur(b,g,a) = 66.062, 340.060, 0.259 ms
HTTP 200: {"choices":[{"finish_reason":"length", ..., "reasoning_content":"Here's a thinking process: ..."}],
  "timings":{"predicted_n":300,"predicted_ms":2910.654,"predicted_per_second":103.07, ...}}
```

### Summary table

| Backend | Build dir | Model | NextN | tg t/s | Accept rate | Output |
|---|---|---|---|---|---|---|
| Vulkan | `build/` | UDT Q4_K_XL | OFF | 118.6 | n/a | coherent |
| Vulkan | `build/` | UDT Q4_K_XL | ON (draft-max 2) | n/a (500) | **0.67%** (2/297) | garbled |
| Vulkan | `build/` | UDT Q4_K_XL | ON (draft-max 1) | n/a (500) | **0.68%** (1/147) | garbled |
| Vulkan | `build/` (with sync patch) | UDT Q4_K_XL | ON (draft-max 2) | n/a (500) | **1.01%** (3/296) | garbled |
| Vulkan | `build/` (`GGML_VK_DISABLE_ASYNC=1`) | UDT Q4_K_XL | ON (draft-max 2) | n/a (500) | **0.67%** (2/297) | garbled |
| Vulkan | `build/` | unsloth UD-Q4_K_M | ON (draft-max 2) | n/a (500) | **1.01%** (2/197) | garbled |
| HIP | `build-hip/` | UDT Q4_K_XL | OFF | 76.0 | n/a | coherent |
| HIP | `build-hip/` | UDT Q4_K_XL | ON (draft-max 2) | **103.1** | **81.21%** (134/165) | **coherent, +35.7% vs OFF** |

The +35.7% HIP gain matches the M4 Max +36% extrapolation in README §0 / NEXTN.md §7, confirming the NextN implementation itself is correct end-to-end — the bug is localized to the Vulkan backend's NextN draft execution path. The model GGUF (UDT mask vs upstream unsloth UD) does not matter — both reproduce identically on Vulkan.

### Hypotheses (in order of likelihood)

Localizing to the Vulkan backend, the most likely culprits are:

- **H1 (most likely after experiment)**: Vulkan kernels for the NextN draft graph (`llm_build_qwen35moe_nextn` and adjacent kernels for the NextN-specific tensors `model.layers[n_main + i].nextn.*`) produce numerically incorrect outputs on RADV, so the draft logits are uncorrelated with the target's. The fact that we get ~0.67-0.68% rather than uniform-random (1/152K vocab ≈ 6.5e-6) suggests *some* correlated signal but degraded — possibly only a subset of NextN tensors are mis-computed on Vulkan.
- **H2 (RULED OUT by experiment)**: `t_h_pre_norm` extraction or the `embd_pre_norm` CPU buffer transfer being async-racey on Vulkan. See "Patch attempt" below — explicit `ggml_backend_sched_synchronize(sched.get())` after the get_async at `src/llama-context.cpp:1666` and `:2111` did not fix the bug; `GGML_VK_DISABLE_ASYNC=1` (which routes all transfers through synchronous path) also did not fix it. Therefore the bug is in actual GPU compute, not buffer transfer / sync.
- **H3**: `cparams.nextn_draft = true` + `kv_only_nextn = true` override in `llama_context` ctor is mis-honored on Vulkan, leaving the draft KV cache uninitialized/wrong-sized.

H1 is most consistent with "draft engages, accept very low" plus "target output is also garbled" — if the Vulkan-computed `t_h_pre_norm` tensor (or any NextN-graph tensor) has numerically wrong values, both the draft predictions and the target's sampled tokens (which read back from the same buffer) would suffer.

### Patch attempt (ruled out async transfer race)

Diff applied (then reverted) for the experiment:

```diff
--- a/src/llama-context.cpp
+++ b/src/llama-context.cpp
@@ -1666,6 +1666,8 @@
         ggml_backend_tensor_get_async(backend_h, t_h_pre_norm, embd_pre_norm.data, 0, n_tokens * n_embd_h * sizeof(float));
+        ggml_backend_sched_synchronize(sched.get());
     }
@@ -2111,6 +2113,8 @@
                 ggml_backend_tensor_get_async(backend_h, t_h_pre_norm, embd_h_out, 0, n_outputs * n_embd_h * sizeof(float));
+                ggml_backend_sched_synchronize(sched.get());
             }
```

Result with patch (Vulkan, draft-max 2): `accept rate = 0.01014 (3/296)`. Marginally different from unpatched 0.67% but still essentially random — output still garbled, HTTP 500.

Result with `GGML_VK_DISABLE_ASYNC=1` env var (which sets `vk_backend->iface.get_tensor_async = nullptr` and routes through sync `get_tensor`): `accept rate = 0.00673 (2/297)` — identical to unpatched. Total smoke time 36 s (vs 4 s with async), confirming the env var works as expected (slower sync transfers). Same garbled output, same HTTP 500.

Both experiments rule out the buffer-transfer race hypothesis. The bug must be in the actual Vulkan compute graph for NextN-related tensors.

### Notes

- The Vulkan backend itself is **not broken in general** — the same build produces correct, fast inference (118.6 t/s) when `--spec-type` is omitted on the same GGUF + same device. Only the NextN draft path is affected.
- `--draft-max 1` reproduces the failure with identical accept rate (0.68%), confirming the chain length is not the cause.
- Both AtomicChat UDT and the upstream `unsloth/Qwen3.6-35B-A3B-MTP-GGUF` reproduce identically on Vulkan, so the UDT mask (NextN pinned to Q8_0, attn Q/K lifted to Q6_K) is not the cause.
- Happy to provide full server logs, GGUF metadata dump, or run additional measurements you want. GLSL extension presence at build time: `GL_KHR_cooperative_matrix supported`, `GL_NV_cooperative_matrix2 not supported`, `GL_EXT_integer_dot_product not supported`, `GL_EXT_bfloat16 not supported`.
